smack-experimental/src/main/java/org/jivesoftware/smackx/message_fastening/MessageFasteningManager.java

```diff
@@ -0,0 +1,100 @@
+import org.jivesoftware.smack.XMPPConnection;
+import org.jivesoftware.smack.XMPPConnectionRegistry;
+import org.jivesoftware.smack.packet.Stanza;
+import org.jivesoftware.smackx.disco.ServiceDiscoveryManager;
+import org.jivesoftware.smackx.message_fastening.element.FasteningElement;
+
+/**
+ * Smacks API for XEP-0422: Message Fastening.
+ * The API is still very bare bones, as the XEP intends Message Fastening to be used as a tool by other protocols.
+ *
+ * To fasten a payload to a previous message, create an {@link FasteningElement} using the builder provided by
+ * {@link FasteningElement#builder()}.
+ *
+ * You need to provide the {@link org.jivesoftware.smackx.sid.element.OriginIdElement} of the message you want to reference.
+ * Then add wrapped payloads using {@link FasteningElement.Builder#addWrappedPayloads(List)}
+ * and external payloads using {@link FasteningElement.Builder#addExternalPayloads(List)}.
+ *
+ * If you fastened some payloads onto the message previously and now want to replace the previous fastening, call
+ * {@link FasteningElement.Builder#isRemovingElement()}.
+ * Once you are finished, build the {@link FasteningElement} using {@link FasteningElement.Builder#build()} and add it to
+ * a stanza by calling {@link MessageFasteningManager#addFasteningElementToStanza(FasteningElement, Stanza)}.
+ *
+ * @see <a href="https://xmpp.org/extensions/xep-0422.html">XEP-0422: Message Fastening</a>
+ */
+public final class MessageFasteningManager extends Manager {
+
+    public static final String NAMESPACE = "urn:xmpp:fasten:0";
+
+    private static final WeakHashMap<XMPPConnection, MessageFasteningManager> INSTANCES = new WeakHashMap<>();
+
+    static {
+        XMPPConnectionRegistry.addConnectionCreationListener(new ConnectionCreationListener() {
+            @Override
+            public void connectionCreated(XMPPConnection connection) {
+                MessageFasteningManager.getInstanceFor(connection).announceSupport();
+            }
+        });
+    }
+
+    private MessageFasteningManager(XMPPConnection connection) {
+        super(connection);
+    }
+
+    public static synchronized MessageFasteningManager getInstanceFor(XMPPConnection connection) {
+        MessageFasteningManager manager = INSTANCES.get(connection);
+        if (manager == null) {
+            manager = new MessageFasteningManager(connection);
+            INSTANCES.put(connection, manager);
+        }
+        return manager;
+    }
+
+    /**
+     * Add the provided Message Fastening apply-to element to the provided stanza.
+     * Note: The stanza MUST NOT contain more than one apply-to elements at the same time.
+     *
+     * @see <a href="https://xmpp.org/extensions/xep-0422.html#rules">XEP-0422 §4: Business Rules</a>
+     *
+     * @param applyTo message fastening apply-to element
+     * @param stanza stanza
+     */
+    public void addFasteningElementToStanza(FasteningElement applyTo, Stanza stanza) {
```

Comment:
That method can be static. And I also think it potentially belongs into `FasteningElement`, since it is not really a Manager type of operation that it performs. It is more like a "modify stanza operation". That said, the type of the second parameter should be a builder. And I wonder if we should restrict it to message type stanzas, i.e. `MessageBuilder`.

Reply:
So maybe add a non static method `FasteningElement.addTo(MessageBuilder)`?